Source/Pawnmorphs/Esoteria/Genebank/ChamberDatabase.cs

```diff
@@ -0,0 +1,423 @@
+			if (FreeStorage < requiredStorage)
+			{
+				reason = NOT_ENOUGH_STORAGE_REASON.Translate(entry.GetCaption(), DatabaseUtilities.GetStorageString(requiredStorage), DatabaseUtilities.GetStorageString(FreeStorage));
+				return false;
+			}
+
+			if (!CanTag)
+			{
+				reason = NOT_ENOUGH_POWER.Translate();
+				return false;
+			}
+
+            if (entry.CanAddToDatabase(this, out reason) == false)
+            {
+                return false;
+            }
+
+			reason = "";
+			return true;
+		}
+
+
+
+		/// <summary>
+		///     Determines whether any of the specified mutation definitions can be
+		///     added to the database, and outputs an error if not.
+		/// </summary>
+		/// <param name="mutationDefs">The mutation definitions.</param>
+		/// <param name="reason">The reason the mutation cannot be ad.</param>
+		/// <returns>
+		///     <c>true</c> if at least one mutation definition can be added to database, otherwise <c>false</c>.
+		/// </returns>
+		public bool CanAddAnyToDatabase([NotNull] IEnumerable<MutationDef> mutationDefs, out string reason)
+		{
+			if (mutationDefs == null) throw new ArgumentNullException(nameof(mutationDefs));
+
+            MutationGenebankEntry entry = new MutationGenebankEntry(mutationDefs.First());
+            _genebankDatabase.TryGetValue(typeof(GenebankEntry<MutationDef>), out ExposedListContainer<IGenebankEntry> mutationEntries);
+            int? minRequiredCapacity = null;
+            MutationDef smallestMutation = null;
+			int neededCapacity = 0;
+            if (mutationEntries != null)
+            {
+                bool hasValid = false;
+                foreach (MutationDef mutationDef in mutationDefs)
+                {
+                    entry.Value = mutationDef;
+                    if (mutationEntries.Contains(entry) == false)
+                    {
+                        hasValid = true;
+                        neededCapacity = entry.GetRequiredStorage();
+                        if (minRequiredCapacity == null || minRequiredCapacity > neededCapacity)
+                        {
+                            minRequiredCapacity = neededCapacity;
+                            smallestMutation = mutationDef;
+						}
+                    }
+				}
+
+                if (hasValid == false)
+                {
+				    // All already tagged.
+				    reason = ALREADY_TAGGED_MULTI_REASON.Translate();
+				    return true;
+                }
+            }
+
+			if (FreeStorage < minRequiredCapacity)
+			{
+				reason = NOT_ENOUGH_STORAGE_REASON.Translate(smallestMutation, DatabaseUtilities.GetStorageString(smallestMutation.GetRequiredStorage()), DatabaseUtilities.GetStorageString(FreeStorage));
+				return false;
+			}
+
+			if (!CanTag)
+			{
+				reason = NOT_ENOUGH_POWER.Translate();
+				return false;
+			}
+
+			reason = "";
+			return true;
+		}
+
+        /// <summary>
+        ///     Exposes the data.
+        /// </summary>
+        public override void ExposeData()
```

Comment:
This method will need to be updated if a new type of genebank entry gets added.  It's worth adding at least a comment to remind people, but it'd be even nicer if it could store all kinds of GenebankEntry types in a generic way (which would also require importing old save data, if it exists)